File: python/sparse_flow.py
import numpy as np
import cv2 as cv
import argparse
from PIL import Image
from time import perf_counter
from utils import FlowSet
    
def main():

    args = argparse.ArgumentParser(
        prog='LK Sparse Optical Flow timer',
        description='Measure the time for Lucas-Kande sparse optical flow.'
    )
    args.add_argument('--plot', action='store_true', help='Whether to plot using OpenCVs plotting functions.')
    args = args.parse_args()
    plot = args.plot

    # params for Shi-Tomasi corner detection
    feature_params = dict(
        maxCorners = 100,
        qualityLevel = 0.3,
        minDistance = 7,
        blockSize = 7 
    )

    # params for lucas kanade optical flow
    lk_params = dict(
        winSize  = (15, 15),
        maxLevel = 2,
        criteria = (cv.TERM_CRITERIA_EPS | cv.TERM_CRITERIA_COUNT, 10, 0.03)
    )

    color = np.random.randint(0, 255, (100, 3))
    data = FlowSet('../data', occ=True)

    start = perf_counter()
    for idx, (img1, img2, _) in enumerate(data):
        img1_gray = cv.cvtColor(img1, cv.COLOR_BGR2GRAY)
        img2_gray = cv.cvtColor(img2, cv.COLOR_BGR2GRAY)

        # corners/features to track flow using ST method
        f0 = cv.goodFeaturesToTrack(img1_gray, mask = None, **feature_params)

        # calculate optical flow
        f1, st, err = cv.calcOpticalFlowPyrLK(img1_gray, img2_gray, f0, None, **lk_params)

        if plot:
            mask = np.zeros_like(img1)
            if f1 is not None:
                good_new = f1[st==1]
                good_old = f0[st==1]
            # draw the tracks
            for i, (new, old) in enumerate(zip(good_new, good_old)):
                a, b = new.ravel()
                c, d = old.ravel()
                mask = cv.line(mask, (int(a), int(b)), (int(c), int(d)), color[i].tolist(), 2)
                img2 = cv.circle(img2, (int(a), int(b)), 5, color[i].tolist(), -1)
        
            img = cv.add(img2, mask)
            cv.imshow('old', img1)
            cv.imshow('frame', img)

        k = cv.waitKey(1) & 0xFF
        if k == 27: # Break if ESC is pressed while plotting
            break

    end = perf_counter()
    print(f"Total time = {end - start:.4f} | Average = {(end - start) / len(data):.4f}")


if __name__ == "__main__":
    main()

File: cpp/sparse_flow.cpp
#include <iostream>
#include "../utils/utils.h"
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/video.hpp>
#include <chrono>
#include <vector>

int main(int argc, char** argv) {
    // === Command line (kept only the plot flag) ===
    cv::CommandLineParser parser(argc, argv,
        "{ h help |      | print this help message }"
        "{ @plot  |      | show optical flow tracks }");
    parser.about("Fast Lucas-Kanade on KITTI 2012/2015");
    if (parser.has("help")) {
        parser.printMessage();
        return 0;
    }
    bool plot = parser.has("@plot");

    // Random colors for visualization
    std::vector<cv::Scalar> colors;
    colors.reserve(100);
    cv::RNG rng;
    for (int i = 0; i < 100; ++i) {
        colors.emplace_back(rng.uniform(0, 256), rng.uniform(0, 256), rng.uniform(0, 256));
    }

    FlowSet dataset("../../data", true);

    cv::Mat mask;                     // reused if plotting

    cv::Mat img1_gray, img2_gray;
    std::vector<cv::Point2f> prevPts, nextPts;
    std::vector<uchar> status;
    std::vector<float> err;

    auto start = std::chrono::high_resolution_clock::now();
    for (size_t i = 0; i < dataset.size(); ++i) {
        auto sample = dataset.get(i);

        cv::cvtColor(sample.img1, img1_gray, cv::COLOR_BGR2GRAY);
        cv::cvtColor(sample.img2, img2_gray, cv::COLOR_BGR2GRAY);

        // Reset mask every frame (pairs are independent!)
        if (plot) {
            mask = cv::Mat::zeros(sample.img1.size(), sample.img1.type());
        }

        // Detect features only on first image
        prevPts.clear();
        cv::goodFeaturesToTrack(img1_gray, prevPts, 100, 0.3, 7, cv::Mat(), 7, false, 0.04);

        if (prevPts.empty()) continue;

        // Prepare nextPts (let LK start from prevPts positions)
        nextPts = prevPts;

        cv::TermCriteria criteria(cv::TermCriteria::COUNT + cv::TermCriteria::EPS, 10, 0.03);
        cv::calcOpticalFlowPyrLK(img1_gray, img2_gray,
                                 prevPts, nextPts, status, err,
                                 cv::Size(15, 15), 2, criteria);

        // Build good_new + draw (only if plotting)
        if (plot) {
            for (size_t j = 0; j < prevPts.size(); ++j) {
                if (status[j] == 1) {
                    cv::line(mask, nextPts[j], prevPts[j], colors[j], 2);
                    cv::circle(sample.img2, nextPts[j], 5, colors[j], -1);  // note: sample.img2 is const ref but we don't modify the stored one
                }
            }
            cv::Mat display;
            cv::add(sample.img2, mask, display);
            cv::imshow("Optical Flow", display);
            int k = cv::waitKey(1);
            if (k == 'q' || k == 27) break;
        }
    }

    auto stop = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(stop - start);
    std::cout << "Time running: " << duration.count() / 1e6 << " seconds\n";

    return 0;
}